llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -2394,6 +2395,11 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
       if (!APF.isZero() && TLI->has(Func))
         return ConstantFoldFP(logb, APF, Ty);
       break;
+    case LibFunc_log1p:
+    case LibFunc_log1pf:
+      if (APF > APFloat(APF.getSemantics(), "-1") && TLI->has(Func))
```

Comment:
Avoid the string? Also still fold if call is memory none

Reply:
There is an APFloat::getOne constructor

Reply:
What do you mean doesn't work?

Reply:
`IEEEFloat::IEEEFloat(const fltSemantics &ourSemantics, integerPart value)` doesn't support initialize itself with a negative value:

https://github.com/llvm/llvm-project/blob/81fee740d073640c79c0d45a8e8e804c1c5fea40/llvm/lib/Support/APFloat.cpp#L1197-L1205
https://github.com/llvm/llvm-project/blob/a31e834ba8beaeed411b85af51ebd6598cb6007c/llvm/include/llvm/ADT/APFloat.h#L1045-L1050

Reply:
So APFloat bug, cool

Reply:
cc @davemgreen Can you have a look and add some unittests?

Reply:
I will fix this.

Reply:
Fixed. Please rebase this patch on https://github.com/llvm/llvm-project/commit/a54d88f97dfeb35e4f6c14e5563ec686e68fd244.